src/main/java/io/strimzi/kafka/bridge/mqtt/core/MqttKafkaMapper.java

```diff
@@ -0,0 +1,161 @@
+
+    // identifies a single level wildcard character in the mqtt pattern. E.g. sensors/+/data
+    private static final String MQTT_TOPIC_SINGLE_LEVEL_WILDCARD_CHARACTER = "+";
+
+    // Regex expression used to replace the + in the mqtt pattern.
+    private static final String SINGLE_LEVEL_WILDCARD_REGEX = "[^/]+";
+
+    // identifies a multi level wildcard character in the mqtt pattern. E.g. sensors/#
+    private static final String MQTT_TOPIC_MULTI_LEVEL_WILDCARD_CHARACTER = "#";
+
+    // matches any character after the string. Used to replace the # in the mqtt pattern.
+    private static final String MULTIPLE_LEVEL_WILDCARD_REGEX = ".*";
+    private final List<MappingRule> rules;
+    private final List<Pattern> patterns = new ArrayList<>();
+    private Pattern placeholderPattern;
+
+    /**
+     * Constructor
+     * <p>
+     * Creates a new instance of MqttKafkaMapper.
+     */
+    public MqttKafkaMapper(List<MappingRule> rules) {
+        this.rules = rules;
+        this.rules.sort(Comparator.comparing(MappingRule::getMqttTopicPatternLevels).reversed());
+        buildRegex();
+    }
+
+    /**
+     * Maps an MQTT topic to a Kafka topic. The topic is mapped according to the defined mapping rules.
+     *
+     * @param mqttTopic represents an MQTT topic.
+     * @return a valid Kafka topic.
+     * @see MappingRule
+     */
+    public String map(String mqttTopic) {
+        for (MappingRule rule : this.rules) {
+            Matcher matcher = this.patterns.get(this.rules.indexOf(rule)).matcher(mqttTopic);
+
+            if (matcher.matches()) {
+                HashMap<String, String> placeholders = new HashMap<>();
+                String mappedKafkaTopic = rule.getKafkaTopicTemplate();
+                String[] mqttTopicPatternParts = rule.getMqttTopicPattern().split("/");
+
+                // find MQTT_TOPIC_PLACEHOLDER_REGEX in the kafkaTopicTemplate.
+                Matcher placeholderMatcher = placeholderPattern.matcher(rule.getKafkaTopicTemplate());
+                while (placeholderMatcher.find()) {
+                    placeholders.put(placeholderMatcher.group(), null);
+                }
+
+                if (!placeholders.isEmpty()) {
+                    for (String placeholderKey : mqttTopicPatternParts) {
+                        if (placeholderKey.matches(MQTT_TOPIC_PLACEHOLDER_REGEX) && placeholders.containsKey(placeholderKey)) {
+                            placeholders.put(placeholderKey, matcher.group(removeBrackets(placeholderKey)));
+                        }
+                    }
+
+                    for (Map.Entry<String, String> entry : placeholders.entrySet()) {
+                        if (entry.getValue() != null) {
+                            mappedKafkaTopic = mappedKafkaTopic.replace(entry.getKey(), entry.getValue());
+                        } else {
+                            throw new IllegalArgumentException("One or more placeholders were not assigned any value.");
```

Comment:
We can be precise about which entry doesn't have a value. Of course, there could be more but because we are throwing an exception and failing, we are not moving next (I mean we are not collecting ALL the entries without a value). So I would say let's be specific in the exception message to show which placeholder didn't get a value so the user knows exactly what to fix.